conda packages

undingen opened the pull request:
__Needs changes:__
- [x] replace 'python'
- [x] remove pip (should be own package)
- [x] figure out how to package all the 3rd party libraries (python abi string)
- [ ] fix package dependencies - are they correct? compare with cpython dependencies
- [ ] do all the builtin modules get build? compare with cpython conda package
- [ ] check if sysconfig is correct
- [ ] run tests during build - maybe we need to remove the tests from the final package
- [ ] strip binaries - I think they are not stripped now that we don't use dpkg to build
- [ ] generate correct .pyc files
- [ ] switch to 'release' from 'opt' build - maybe blocked by instrumentation problems in bolt
- [x] fix test failures during profiling ("file was modified... default.profraw")

__May need changes:__
- [ ] we internally pip install packages into a temporary environment for the bolt profiling task. I think this is okay because this package will not end up in the final conda package and we install them via exact version string but may still be preferred to supply the source packages locally instead.
- [x] ~bolt pkg directly loads source archive from github because we can only checkout tags. Maybe we should fork it and create tags ourselves instead. I named the bolt release we are using after the date of the last commit~ Forked bolt into the pyston org and created a tag which we point to.
 
__How to try it out:__
run the following script to build our packages in a docker miniconda image which outputs the finished packages into `release/conda_pkgs/`
```
$ pyston/conda/build_pkgs.sh
```
Afterwards index the directory and create a activate new enviroment called pyston_env
```
$ conda index -c pyston release/conda_pkgs/
$ conda create -n pyston_env pyston -c `pwd`/release/conda_pkgs/ -c conda-forge --override-channels
$ conda activate pyston_env
```
Now one can e.g install attrs package:
```
(pyston_env)$ conda install --override-channels -c conda-forge -c `pwd`/release/conda_pkgs/ attrs
(pyston_env)$ python -c "import attr"
```
Numpy:
```
(pyston_env)$ conda install --override-channels -c conda-forge -c `pwd`/release/conda_pkgs/ numpy
(pyston_env)$ python -c "import numpy as np; print(np.identity(5))"
```
__Note:__
I used conda-forge instead of the default anaconda channel because conda-forge packages have the python_abi tag set -
on the main channel it will happily let you install cpython extensions compiled against cpython.


__Note:__
The `compiler-rt` recipe is from anaconda but they currently don't build a linux_64 version thats why we also have compile it for now and I included it.

undingen committed:
conda: build first packages

undingen committed:
add option to use prebuild binaries of llvm and bolt
Setting PYSTON_USE_SYS_BINS=1 will make us use prebuild binaries.
The user must make sure it has the required version installed (llvm 10)

Prebuild packages don't normally supply the perf jit events so disabled it for out build too.
We don't need it anymore because nitrous is not a JIT anymore.

undingen committed:
conda: don't bundle pip

undingen committed:
conda: create python replacement package

kmod commented on pyston/conda/build_pkgs.sh:
```diff
@@ -0,0 +1,23 @@
+conda build pyston/conda/pyston
+conda install pyston --use-local
+
+wget 
```
Is this a leftover?

undingen replied:
I accidentally added that file :blush:.

kmod reviewed (commented):
This is looking great!

undingen committed:
conda: enable PGO
Before we did not actually use profiling because configure is checking $CC for gcc but it was called '_cc'.
Copied in some code from the offical cpython conda recipe which fixes this.
This brought up two new errors which I had to fix.

This also fixes the test failures during profiling ("file was modified... default.profraw")

undingen committed:
switch bolt repo to pyston/bolt
This way we can create tags for the version pyston is using.

undingen committed:
conda: create python_abi package
conda-forge uses this to support pypy and cpython
we use the same idea to also support pyston

undingen committed:
conda: change pyston site-package to cpythons directory
this is necessary otherwise conda install noarch python packages can't be imported
because they will get placed in the wrong directory.

undingen committed:
conda: don't hardcode version numbers in scripts

undingen committed:
conda: split into pyston2.3 and pyston package
rename the package previously called pyston to pyston2.3 (which contains the actual interpreter)
and introduce new noarch pyston package which just installs the the pyston python package.
This is very similar to how the pypy recipes work

this makes it possible to create a new pyston environment using:
conda create -n pyston_env pyston
instead of the more complicated
conda create -n pyston_env python="3.8.*=*_pyston"

undingen committed:
conda: add python3.8 symlink

undingen committed:
conda: remove all site-packages

undingen committed:
fix build inside miniconda docker

undingen committed:
conda: always create lib/python3.8/site-packages
Include our lib/python3.8/site-packages/README.txt to make sure
this directory always exist.
Otherwise 3rd party conda packages will fail during install.

undingen committed:
conda: add script which builds conda packages

undingen committed:
nitrous: define __STDC_FORMAT_MACROS
this fixes a build error I was seeing via conda:
/opt/conda/conda-bld/pyston2.3_1633957270008/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/include/llvm/Object/SymbolicFile.h:47:31: error: expected ')'
  OS << "(" << format("0x%08" PRIxPTR, D.p) << " (" << format("0x%08x", D.d.a)

Same fix as here:
https://github.com/openai/triton/commit/7fdf2e378c1ce2a285e4d9c0d77201735836336a

undingen committed:
conda: add python_abi, reset build_num and fix build_pkgs script

undingen committed:
conda: build numpy package
